AX: HTML menu element should map to role=list

SirriCelles opened the pull request:
#### f816c098797e3849e3871f03c30de2a850c59168
<pre>
2022-09-05 SirriCelles &lt;sirricelles@gmail.com&gt;

    AX: HTML menu element should map to role=list
    https://bugs.webkit.org/show_bug.cgi?id=201581
</pre>
----------------------------------------------------------------------
#### 2d2c29446b11ee9e8e0ff4e8ba64dd761a053d43
<pre>
2022-09-02 SirriCelles &lt;sirricelles@gmail.com&gt;
    update LayoutTests

    Tests:
    * LayoutTests/platform/mac-wk1/accessibility/roles-exposed-expected.txt
    * LayoutTests/platform/mac-wk2/accessibility/roles-exposed-expected.txt
    * LayoutTests/platform/win/accessibility/roles-exposed-expected.txt
    * LayoutTests/platform/wpe/accessibility/roles-exposed-expected.txt
</pre>
----------------------------------------------------------------------
#### f9011a0dcb12f958e55f15c4f2c162b63df54667
<pre>
2022-09-02 SirriCelles &lt;sirricelles@gmail.com&gt;

    Test:
       LayoutTests/platform/glib/accessibility/aria-visible-element-roles-expected.txt
       LayoutTests/platform/win/accessibility/roles-exposed-expected.txt
       LayoutTests/platform/wpe/accessibility/aria-visible-element-roles-expected.txt
       LayoutTests/platform/wpe/accessibility/roles-exposed-expected.txt
</pre>
----------------------------------------------------------------------
#### eb21d127add74ea6d55b63c83370c49121929e41
<pre>
2022-09-02 SirriCelles &lt;sirricelles@gmail.com&gt;

Tests: LayoutTests/platform/mac-wk1/accessibility/roles-exposed-expected.txt
         LayoutTests/platform/mac-wk2/accessibility/roles-exposed-expected.txt
</pre>
----------------------------------------------------------------------
#### a5792c113dc2621404e7600250ac4a3e12e1ebea
<pre>
2022-09-1 SirriCelles &lt;sirricelles@gmail.com&gt;

    AX: HTML menu element should map to role=list
    <a href="https://bugs.webkit.org/show_bug.cgi?id=201581">https://bugs.webkit.org/show_bug.cgi?id=201581</a>

    This patch modifies Source/WebCore/accessibility/AccessibilityNodeObject.cpp file.
    Modified check that returns AccessibilityRole::Toolbar for menu element with attribute=&quot;toolbar&quot; and AccessibilityRole::rolelist
    for the menu element. Per the HTML AAM, &lt;menu&gt; is mapped to list role but we keep backward compatibility for the obsolete type=&quot;toolbar&quot;.

    Tests:  LayoutTests/accessibility/roles-exposed.html

    *LayoutTests/accessibility/roles-exposed-expected.txt
        *LayoutTests/accessibility/roles-exposed.html
        *LayoutTests/platform/mac-wk1/accessibility/aria-visible-element-roles-expected.txt
        *LayoutTests/platform/mac-wk1/accessibility/roles-exposed-expected.txt
        *LayoutTests/platform/mac-wk2/accessibility/roles-exposed-expected.txt
        *LayoutTests/platform/win/accessibility/aria-visible-element-roles-expected.txt
        *LayoutTests/platform/win/accessibility/roles-exposed-expected.txt
    *Source/WebCore/accessibility/AccessibilityNodeObject.cpp
</pre>
----------------------------------------------------------------------
#### 7d9f43b71ceb14a0dbe913a0919015a6cd2e6132
<pre>
2022-08-23 SirriCelles &lt;sirricelles@gmail.com&gt;

    AX: HTML menu element should map to role=list
    <a href="https://bugs.webkit.org/show_bug.cgi?id=201581">https://bugs.webkit.org/show_bug.cgi?id=201581</a>

    This patch modifies Source/WebCore/accessibility/AccessibilityNodeObject.cpp file.
    Modified check that returns AccessibilityRole::Toolbar for menu element with attribute=&quot;toolbar&quot; and AccessibilityRole::rolelist
    for the menu element. Restructure check
</pre>
----------------------------------------------------------------------
#### 7a48c965b49e17cd90c7976b6f90cf794a5f9c9e
<pre>
2022-08-23 SirriCelles &lt;sirricelles@gmail.com&gt;

    AX: HTML menu element should map to role=list
    <a href="https://bugs.webkit.org/show_bug.cgi?id=201581">https://bugs.webkit.org/show_bug.cgi?id=201581</a>

    This patch modifies Source/WebCore/accessibility/AccessibilityNodeObject.cpp file.
    Modified check that returns AccessibilityRole::Toolbar for menu element with attribute=&quot;toolbar&quot; and AccessibilityRole::rolelist
    for the menu element.
</pre>
----------------------------------------------------------------------
#### 3f22c0338e8af6f2634d6ac092f821d5fff09a30
<pre>
2022-08-23 SirriCelles &lt;sirricelles@gmail.com&gt;

    AX: HTML menu element should map to role=list
    <a href="https://bugs.webkit.org/show_bug.cgi?id=201581">https://bugs.webkit.org/show_bug.cgi?id=201581</a>

    This patch modifies Source/WebCore/accessibility/AccessibilityNodeObject.cpp file.
    <a href="https://github.com/WebKit/WebKit/blob/main/Source/WebCore/accessibility/AccessibilityNodeObject.cpp#L422.">https://github.com/WebKit/WebKit/blob/main/Source/WebCore/accessibility/AccessibilityNodeObject.cpp#L422.</a>
    It adds check that returns AccessibilityRole::Toolbar for menu element with attribute=&quot;toolbar&quot; and AccessibilityRole::rolelist
    for the menu element.

    Test: LayoutTests/accessibility/aria-visible-element-roles.html

    *  Source/WebCore/accessibility/AccessibilityNodeObject.cpp
    *  LayoutTests/accessibility/aria-visible-element-roles.html
    *  LayoutTests/platform/mac-wk2/accessibility/aria-visible-element-roles-expected.txt
</pre>
----------------------------------------------------------------------
#### 988c8c64467f0f0d80eb9c7a997263673f316562
<pre>
2022-08-23 SirriCelles &lt;sirricelles@gmail.com&gt;

    AX: HTML menu element should map to role=list
    <a href="https://bugs.webkit.org/show_bug.cgi?id=201581">https://bugs.webkit.org/show_bug.cgi?id=201581</a>

    This patch modifies Source/WebCore/accessibility/AccessibilityNodeObject.cpp file.
    <a href="https://github.com/WebKit/WebKit/blob/main/Source/WebCore/accessibility/AccessibilityNodeObject.cpp#L422.">https://github.com/WebKit/WebKit/blob/main/Source/WebCore/accessibility/AccessibilityNodeObject.cpp#L422.</a>
    It adds check that returns AccessibilityRole::Toolbar for menu element with attribute=&quot;toolbar&quot; and AccessibilityRole::rolelist
    for the menu element.

    Test: LayoutTests/accessibility/aria-visible-element-roles.html
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f816c098797e3849e3871f03c30de2a850c59168

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/88302 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/32677 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/19109 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/97497 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/152963 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/92269 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/31177 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/26858 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/80498 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/92153 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93911 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/24840 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/75105 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/24813 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/79747 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79867 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/67779 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/28761 "Built successfully") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/13839 "Found 1 new test failure: accessibility/roles-exposed.html (failure)") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/28762 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14847 "Found 1 new test failure: accessibility/roles-exposed.html (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/31917 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/37756 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/30915 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33985 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

fred-wang commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -419,8 +419,12 @@ AccessibilityRole AccessibilityNodeObject::determineAccessibilityRoleFromNode(Tr
+    if (node()->hasTagName(menuTag) && equalLettersIgnoringASCIICase(getAttribute(typeAttr), "toolbar"_s)) {
         return AccessibilityRole::Toolbar;
+    } else 
+        return AccessibilityRole::List;
```
This does not look correct. It will map any `<menu type="toolbar">` to Toolbar and any other node (whatever their tag name) to a List.

(I wonder if we want to keep current mapping for `<menu type="toolbar">` as Firefox/Chromium don't do that AFAIK, but I'll let a11y reviewers comment on that)

fred-wang replied:
BTW, this is the result of `Tools/Scripts/check-webkit-style` by EWS:

ERROR: Source/WebCore/accessibility/AccessibilityNodeObject.cpp:424:  One line control clauses should not use braces.  [whitespace/braces] [4]
ERROR: Source/WebCore/accessibility/AccessibilityNodeObject.cpp:422:  An else statement can be removed when the prior "if" concludes with a return, break, continue or goto statement.  [readability/control_flow] [4]

SirriCelles replied:
Ok That's true, Let me modify the code to fix this.

SirriCelles replied:
@fred-wang  I have updated the PR with the style checks passing now.

fred-wang commented on LayoutTests/accessibility/aria-visible-element-roles.html:
```diff
@@ -36,6 +36,10 @@
 
+    <menu hidden aria-hidden="false" id="menu-list">
+        <li hidden aria-hidden="false" id="li-element">Hello world</li>
+    </menu>
```
Great, this verifies that menu without type="toolbar" indeed maps to a string role and we have the case with an explicit type already tested.

However, I wonder why the case for `<hr>` and `<time>` are not broken by your patch (see my comment on the C++ code below). Were you able to run the test?

fred-wang commented on LayoutTests/platform/mac-wk2/accessibility/aria-visible-element-roles-expected.txt:
```diff
@@ -120,6 +120,10 @@ This test ensures ARIA visible elements (e.g. those with both `hidden` and `aria
+<menu hidden="" aria-hidden="false" id="menu-list"></menu>
+    AXRole: AXList
+    computedRoleString: list
+
```
Probably we need that for other platforms. Let's see what EWS outputs...

fred-wang replied:
It seems you need to update the gtk and win expectations too. You can grab the results from the build bots.

SirriCelles replied:
Ok. How do I do that?

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -421,6 +421,8 @@ AccessibilityRole AccessibilityNodeObject::determineAccessibilityRoleFromNode(Tr
     // menu tags with toolbar type should have Toolbar role.
     if (node()->hasTagName(menuTag) && equalLettersIgnoringASCIICase(getAttribute(typeAttr), "toolbar"_s))
         return AccessibilityRole::Toolbar;
+    if (node()->hasTagName(menuTag) && !equalLettersIgnoringASCIICase(getAttribute(typeAttr), "toolbar"_s))
```
Rather than checking for the menu tagname twice, could we structure it differently to do less work? Maybe something like this?

```
if (node()->hasTagName(menuTag)) {
    if (equalLettersIgnoringASCIICase(getAttribute(typeAttr), "toolbar"_s))
        return AccessibilityRole::Toolbar;
    return AccessibilityRole::List;
}

SirriCelles replied:
Perfect. I wasn't sure how to structure this. Thanks, @twilco. Let me make an upate

fred-wang replied:
Was about to suggest the same thing. This looks good to me now. Although perhaps we want to add a C++ comment about why we are not following the spec here e.g.

    // Per the HTML AAM, <menu> is mapped to list role but we keep backward compatibility for the obsolete type="toolbar".
    // See https://html.spec.whatwg.org/multipage/obsolete.html#attr-menu-type

fred-wang commented on LayoutTests/accessibility/roles-exposed.html:
```diff
@@ -218,6 +218,8 @@ <h2 id="section-label">X</h2>
     <li data-platform="atspi,mac" class="ex">X</li>
 </ul>
+
+
```
Why this whitespace change? I guess you mean to add a test for menu here as suggested by @twilco?

SirriCelles replied:
Yes. I must have left this in.

fred-wang commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -418,9 +418,12 @@ AccessibilityRole AccessibilityNodeObject::determineAccessibilityRoleFromNode(Tr
-    // menu tags with toolbar type should have Toolbar role.
-    if (node()->hasTagName(menuTag) && equalLettersIgnoringASCIICase(getAttribute(typeAttr), "toolbar"_s))
-        return AccessibilityRole::Toolbar;
+    // menu tags with toolbar type should have Toolbar role and menu tags without toolbar should have role list
```
I guess explain that the HTML AAM spec says to always map to role list and that special handling for type="toolbar" is only for backward compatibility.

SirriCelles committed:
2022-08-23 SirriCelles <sirricelles@gmail.com>

    AX: HTML menu element should map to role=list
    https://bugs.webkit.org/show_bug.cgi?id=201581

    This patch modifies Source/WebCore/accessibility/AccessibilityNodeObject.cpp file.
    https://github.com/WebKit/WebKit/blob/main/Source/WebCore/accessibility/AccessibilityNodeObject.cpp#L422.
    It adds check that returns AccessibilityRole::Toolbar for menu element with attribute="toolbar" and AccessibilityRole::rolelist
    for the menu element.

    Test: LayoutTests/accessibility/aria-visible-element-roles.html

    *  Source/WebCore/accessibility/AccessibilityNodeObject.cpp
    *  LayoutTests/accessibility/aria-visible-element-roles.html
    *  LayoutTests/platform/mac-wk2/accessibility/aria-visible-element-roles-expected.txt

SirriCelles committed:
2022-08-23 SirriCelles <sirricelles@gmail.com>

    AX: HTML menu element should map to role=list
    https://bugs.webkit.org/show_bug.cgi?id=201581

    This patch modifies Source/WebCore/accessibility/AccessibilityNodeObject.cpp file.
    Modified check that returns AccessibilityRole::Toolbar for menu element with attribute="toolbar" and AccessibilityRole::rolelist
    for the menu element.

SirriCelles committed:
2022-08-23 SirriCelles <sirricelles@gmail.com>

    AX: HTML menu element should map to role=list
    https://bugs.webkit.org/show_bug.cgi?id=201581

    This patch modifies Source/WebCore/accessibility/AccessibilityNodeObject.cpp file.
    Modified check that returns AccessibilityRole::Toolbar for menu element with attribute="toolbar" and AccessibilityRole::rolelist
    for the menu element. Restructure check

SirriCelles committed:
2022-08-23 SirriCelles <sirricelles@gmail.com>

    AX: HTML menu element should map to role=list
    https://bugs.webkit.org/show_bug.cgi?id=201581

    This patch modifies Source/WebCore/accessibility/AccessibilityNodeObject.cpp file.
    https://github.com/WebKit/WebKit/blob/main/Source/WebCore/accessibility/AccessibilityNodeObject.cpp#L422.
    It adds check that returns AccessibilityRole::Toolbar for menu element with attribute="toolbar" and AccessibilityRole::rolelist
    for the menu element.

    Test: LayoutTests/accessibility/aria-visible-element-roles.html

SirriCelles committed:
2022-09-1 SirriCelles <sirricelles@gmail.com>

    AX: HTML menu element should map to role=list
    https://bugs.webkit.org/show_bug.cgi?id=201581

    This patch modifies Source/WebCore/accessibility/AccessibilityNodeObject.cpp file.
    Modified check that returns AccessibilityRole::Toolbar for menu element with attribute="toolbar" and AccessibilityRole::rolelist
    for the menu element. Per the HTML AAM, <menu> is mapped to list role but we keep backward compatibility for the obsolete type="toolbar".

    Tests:  LayoutTests/accessibility/roles-exposed.html

    *LayoutTests/accessibility/roles-exposed-expected.txt
 	*LayoutTests/accessibility/roles-exposed.html
 	*LayoutTests/platform/mac-wk1/accessibility/aria-visible-element-roles-expected.txt
 	*LayoutTests/platform/mac-wk1/accessibility/roles-exposed-expected.txt
 	*LayoutTests/platform/mac-wk2/accessibility/roles-exposed-expected.txt
 	*LayoutTests/platform/win/accessibility/aria-visible-element-roles-expected.txt
 	*LayoutTests/platform/win/accessibility/roles-exposed-expected.txt
    *Source/WebCore/accessibility/AccessibilityNodeObject.cpp

SirriCelles committed:
2022-09-02 SirriCelles <sirricelles@gmail.com>

Tests: LayoutTests/platform/mac-wk1/accessibility/roles-exposed-expected.txt
    	 LayoutTests/platform/mac-wk2/accessibility/roles-exposed-expected.txt

SirriCelles committed:
2022-09-02 SirriCelles <sirricelles@gmail.com>

    Test:
       LayoutTests/platform/glib/accessibility/aria-visible-element-roles-expected.txt
       LayoutTests/platform/win/accessibility/roles-exposed-expected.txt
       LayoutTests/platform/wpe/accessibility/aria-visible-element-roles-expected.txt
       LayoutTests/platform/wpe/accessibility/roles-exposed-expected.txt

fred-wang commented on LayoutTests/platform/mac-wk1/accessibility/roles-exposed-expected.txt:
```diff
@@ -500,6 +500,16 @@ meter
+      AXRole: AXGroup
+      AXSubrole:
+      AXRoleDescription: group
+
```
Why is this new output placed *after* the output of the meter element above? This does not match the order in `roles-exposed.html`

fred-wang commented on LayoutTests/platform/mac-wk2/accessibility/roles-exposed-expected.txt:
```diff
@@ -500,6 +500,16 @@ meter
+      AXRole: AXGroup
+      AXSubrole:
+      AXRoleDescription: group
+
```
Same here, this does not match the order in `roles-exposed.html`

fred-wang commented on LayoutTests/platform/win/accessibility/roles-exposed-expected.txt:
```diff
@@ -534,6 +534,16 @@ meter
+      AXRole: AXGroup
+      AXSubrole:
+      AXRoleDescription: group
+
```
Same here, this does not match the order in `roles-exposed.html`

fred-wang commented on LayoutTests/platform/wpe/accessibility/roles-exposed-expected.txt:
```diff
@@ -322,6 +322,12 @@ mtr
+
+li
+      AXRole: AXListItem
+
```
The order here is correct! But why didn't you update `LayoutTests/platform/gtk/accessibility/roles-exposed-expected.txt` too?

https://searchfox.org/wubkat/rev/8e1bf48c90b9cc1784acfc301f5cc79d25d622f4/LayoutTests/platform/gtk/accessibility/roles-exposed-expected.txt#324

fred-wang reviewed (changes requested):
It seems the EWS bots are still failing. See comments inline.

SirriCelles committed:
2022-09-02 SirriCelles <sirricelles@gmail.com>
    update LayoutTests

    Tests:
    * LayoutTests/platform/mac-wk1/accessibility/roles-exposed-expected.txt
    * LayoutTests/platform/mac-wk2/accessibility/roles-exposed-expected.txt
    * LayoutTests/platform/win/accessibility/roles-exposed-expected.txt
    * LayoutTests/platform/wpe/accessibility/roles-exposed-expected.txt

SirriCelles committed:
2022-09-05 SirriCelles <sirricelles@gmail.com>

    AX: HTML menu element should map to role=list
    https://bugs.webkit.org/show_bug.cgi?id=201581

fred-wang commented on LayoutTests/accessibility/roles-exposed.html:
```diff
@@ -142,7 +142,9 @@ <h6 data-platform="atspi,mac" class="ex">X</h6>
   </mtable>
 </math>
 </div>
```
Note that the menu is after the MathML elements and before the `<meter>` tag.

fred-wang commented on LayoutTests/platform/gtk/accessibility/roles-exposed-expected.txt:
```diff
@@ -223,6 +223,12 @@ mark
+
+li
+      AXRole: AXListItem
+
```
So I believe it should be placed after all the MathML elements here. I don't see the meter element in the output so maybe it's not up-to-date.

fred-wang commented on LayoutTests/platform/mac-wk1/accessibility/roles-exposed-expected.txt:
```diff
@@ -332,6 +332,16 @@ math
+      AXRole: AXGroup
+      AXSubrole:
+      AXRoleDescription: group
+
```
Same here, it should be after the MathML elements and before the `<meter>`.

fred-wang commented on LayoutTests/platform/mac-wk2/accessibility/roles-exposed-expected.txt:
```diff
@@ -332,6 +332,16 @@ math
+      AXRole: AXGroup
+      AXSubrole:
+      AXRoleDescription: group
+
```
And here again, after the MathML elements and before the `<meter>`

fred-wang commented on LayoutTests/platform/wpe/accessibility/roles-exposed-expected.txt:
```diff
@@ -223,6 +223,12 @@ mark
+
+li
+      AXRole: AXListItem
+
```
Same again here, it should be moved down after the MathML elements.
